LibraBFT/Concrete/Properties/VotesOnce.agda

```diff
@@ -106,274 +106,99 @@ module LibraBFT.Concrete.Properties.VotesOnce where
-    ...| inj₂ refl
-       = sym (PredStep-wlog-ht' preach hip ps v'⊂m' m'∈outs (msg⊆ vOld) (msg∈pool vOld) ver' (msgSigned vOld) hpk (sym e≡) (sym r≡))
-
-    -- (D) Finally, both votes are new in this step. The proof is then trivially
-    -- forwarded to the implementation obligation.
-    PredStep-hh' preach hip ps {v} v⊂m m∈outs v'⊂m' m'∈outs ver ver' hpk e≡ r≡
-       | inj₁ (vValid , vNew) | inj₁ (v'Valid , v'New)
-       = Impl-VO2 preach ps hpk v⊂m m∈outs ver vNew vValid v'⊂m' m'∈outs ver' v'New v'Valid e≡ r≡
-
-    PredStep-hh
-      : ∀{e pid st' outs}{pre : SystemState e}
-      → ReachableSystemState pre
-      → (pstep : StepPeer pre pid st' outs)
-      → Pred pre
-      → ∀{pk v v'}
-      → let pool = msgPool (StepPeer-post pstep)
-         in Meta-Honest-PK pk
-          → (ver  : WithVerSig pk v )(sv  : MsgWithSig∈ pk (ver-signature ver ) pool)
-          → (msgSender sv  , msgWhole sv)  ∈ List-map (pid ,_) outs
-          → (ver' : WithVerSig pk v')(sv' : MsgWithSig∈ pk (ver-signature ver') pool)
-          → (msgSender sv' , msgWhole sv') ∈ List-map (pid ,_) outs
-          → v ^∙ vEpoch ≡ v' ^∙ vEpoch
-          → v ^∙ vRound ≡ v' ^∙ vRound
-          → v ^∙ vProposedId ≡ v' ^∙ vProposedId
-    PredStep-hh preach (step-cheat fm isCheat) hip hpk ver sv (here refl) ver' sv' (here refl) epoch≡ r≡
-      with isCheat (msg⊆ sv) (msgSigned sv)
-    ...| inj₁ abs    = ⊥-elim (hpk abs) -- The key was honest by hypothesis.
-    ...| inj₂ sentb4
-      with isCheat (msg⊆ sv') (msgSigned sv')
-    ...| inj₁ abs     = ⊥-elim (hpk abs) -- The key was honest by hypothesis.
-    ...| inj₂ sentb4'
-      with msgSameSig sv | msgSameSig sv'
-    ...| refl | refl = hip hpk ver sentb4 ver' sentb4' epoch≡ r≡
-    PredStep-hh preach (step-honest x) hip hpk ver sv thisStep ver' sv' thisStep' epoch≡ r≡
-      with sameHonestSig⇒sameVoteData hpk ver (msgSigned sv) (sym (msgSameSig sv))
-    ...| inj₁ abs  = ⊥-elim (meta-sha256-cr abs)
+    -- Case (i) is trivial; cases (iii) and (iv) are symmetric and reduce to an implementation
+    -- obligation (Impl-VO1) and case (ii) reduces to a different implementation obligation
+    -- (Impl-VO2).
+
+
+    VotesOnceProof :
+       ∀ {v v' e pk} {st : SystemState e}
+       → ReachableSystemState st
+       → Meta-Honest-PK pk
+       → (vv  : WithVerSig pk v)  → MsgWithSig∈ pk (ver-signature vv) (msgPool st)
+       → (vv' : WithVerSig pk v') → MsgWithSig∈ pk (ver-signature vv') (msgPool st)
+       → v ^∙ vEpoch ≡ v' ^∙ vEpoch
+       → v ^∙ vRound ≡ v' ^∙ vRound
+       → v ^∙ vProposedId ≡ v' ^∙ vProposedId
+    VotesOnceProof step-0 _ _ msv _ _ _ _ = ⊥-elim (¬Any[] (msg∈pool msv))
+    VotesOnceProof (step-s r (step-epoch _)) pkH vv msv vv' msv' ep≡ r≡
+      = VotesOnceProof r pkH vv msv vv' msv' ep≡ r≡
+    VotesOnceProof (step-s r (step-peer cheat@(step-cheat f c))) pkH vv msv vv' msv' ep≡ r≡
+      with ¬cheatForgeNew cheat refl unit pkH msv | ¬cheatForgeNew cheat refl unit pkH msv'
```

Comment:
I prefer the `w` of `with` to align with the `|` on the following line.  I see we haven't been consistent about that everywhere and of course it's really minor, but in this proof you have three variants in close proximity (once character before, aligned with, and one character after).  It would be nice to make these consistent.

If you feel like it, you could also clarify the `STYLE.md` (line 19) as follows:

>  -  Example `with` abstraction (note alignment of `with` and `|`, and no space between `...` and `|`)'.

Reply:
Ok I will always do that from now on.

Reply:
Done! I will do that from now on.